### PR TITLE
Kruskal's algorithm

### PR DESCRIPTION
**Pull request by cjamadagni**
Implemented Kruskal's minimum spanning tree algorithm.

### CODE REVIEW

File: src/kruskal.jl [hunk 1]
```diff
@@ -0,0 +1,52 @@
+####################################################
+## Minimum Spanning Tree - Kruskal algorithm
+####################################################
+
+"""
+
+Computes the minimum spanning tree of an undirected graph by using Kruskal's algorithm.
+
+Requires arguments:
+distmx::LightGraphs.Graph          # Adjacency matrix of the input graph
```
Comment (sbromberger) [inline src/kruskal.jl]: I'm not sure what this means. distmx is generally a matrix (2x2), not a Graph. Also, we don't need to prefix with LightGraphs here, do we?

File: src/kruskal.jl [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+n::Int                             # The number of vertices in the graph 
+
```
Comment (sbromberger) [inline src/kruskal.jl]: If we have the graph, we don't need the number of vertices. We can get them via `nv()`

File: test/kruskal.jl [hunk 3]
```diff
@@ -0,0 +1,32 @@
+using LightGraphs
+import LightGraphs: kruskal_minimum_spanning_tree
```
Comment (sbromberger) [inline test/kruskal.jl]: Don't need this line here, I don't think.

**Comment by sbromberger** [comment]
Also, please squash commits and make sure you've got conflicts resolved :) thank you!

**Comment by cjamadagni** [comment]
Thank you for the feedback. I'll fix those changes. 

Should i be taking an undirected graph as the function parameter or is it okay to accept an adjacency matrix?

File: src/kruskal.jl [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+Returns: An undirected graph which is the MST of the adjacency matrix.
+
+"""
+
+
+function kruskal_minimum_spanning_tree(distmx, n)
```
Comment (sbromberger) [inline src/kruskal.jl]: Also, please specify types in parameters.

**Comment by sbromberger** [comment]
The standard practice for LG is to accept a `Graph` or `DiGraph` as an argument, but there's nothing that would prevent you from creating both variants: that is, 

```
function kruskal_minimum_spanning_tree{T<:Real}(distmx::Matrix{T})
...

kruskal_minimum_spanning_tree(g::Graph) =  kruskal_minimum_spanning_tree(adjacency_matrix(g))
```

**Comment by CarloLucibello** [comment]
I think this function should also handle sparse graphs efficiently. Therefore I'd like to see a function signature containing an edgemap, something like

```julia
function{T<:Real}(g::Graph, weights::Dict{Edge, T})
. . .
   edgelist = sort(collect(weights), by=x->x[2])
. . . 
end
```

**Comment by sbromberger** [comment]
I am against dicts. We don't use them as arguments anywhere else. If we want sparse matrices then we should use an abstract type that handles both dense and sparse in the argument list.

**Comment by CarloLucibello** [comment]
I used a dict as an argument in the matching algorithm. This seemed to me the most natural structure for storing values associated to edges. 

Should we attempt to define an abstract EdgeMap type, with concrete types for sparse matrixes/non-sparse matrices/dicts/...  storages, and a common interface?

**Comment by sbromberger** [comment]
We should revisit. We're using 

```
distmx::AbstractArray{T, 2}=DefaultDistance();
```

everywhere else.

**Comment by cjamadagni** [comment]
@sbromberger : I've made a few changes. 

```julia
kruskal_minimum_spanning_tree(g::Graph) =  kruskal_minimum_spanning_tree(adjacency_matrix(g))
```

I can't use the above statement since `adjacency_matrix()` doesn't give me the weights between edges. Currently, I am passing a weight matrix.

**Comment by sbromberger** [comment]
@cjamadagni you can still keep that as a type signature for cases where you just want to assume the default distances of 1. It doesn't hurt anything :)

**Comment by cjamadagni** [comment]
@sbromberger : Sorry for the delay. Added that type signature.

**Comment by jpfairbanks** [comment]
So this code is only (asymptotically) efficient for dense graphs. It looks like there need to be two changes for this code to be practical on sparse graphs.

- [ ] Change the iteration over the matrix when constructing the edge list to iterate over edges and lookup matrix elements.
- [ ] change is_cyclic to avoid DFS at each step. We can ensure the cycle free properties of trees at each step without performing the DFS. If you are adding the edge v~u to mst, and v and u are already in the tree, then v~u will introduce a cycle. Otherwise, if v or u is not in the tree, then adding v~u to the graph will not introduce a cycle. So by adding in that check, we can remove the test for is_cyclic, which is O(n) called n-1 times.

**Comment by jpfairbanks** [comment]
@sbromberger, Philosophical question: do we want to enforce constraints like the distance matrix has to be symmetric? That is true if you want it represent real distances, but the algorithm should work just fine without the constraint you just might not get what you need for the application.

There are two goals here: simple code, quality code. For simplicity we should leave out the checks and if it works it works, but for quality making checks to enforce that our inputs match our assumptions will reduce user errors. It comes at the cost of finding some users complaining that we assert a property that should not be enforced.

**Comment by sbromberger** [comment]
@jpfairbanks that's a good question. I'm not entirely opposed to not enforcing symmetry (though we've done it in a few places, but there I think it's ok). The reason is that I can envision a scenario where you'd want a weight from `u` to `v` to be different than the weight from `v` to `u`. Not sure what the implications are for the primitive operations, or whether they'd actually work, though.

**Comment by jpfairbanks** [comment]
On closer inspection, in order to avoid checking is_cyclic on every insertion into the tree, one needs to use a fancier datastructure. My previous comment forgot to handle the hard case where u and v are both in the tree but are not in the same connected component.

We can import IntDisjointSets from DataStructures.jl and use that for the correct implementation of Kruskal's algorithm. I'll post that version of the code soon. But it will add a dependency on DataStructures.jl

**Comment by sbromberger** [comment]
No more dependencies, please. I'm trying to get rid of the ones we have. :)

**Comment by sbromberger** [comment]
@cjamadagni @jpfairbanks what are the plans with this PR? Do we go to review, or do we close it?

**Comment by cjamadagni** [comment]
@sbromberger Since you don't want another dependency, how could we proceed?

**Comment by sbromberger** [comment]
@cjamadagni I'm not seeing any additional dependencies listed here - what am I missing?

**Comment by jpfairbanks** [comment]
It requires DataStructures.jl for the union find structure,right? I think
depending on DataStructures.jl is OK because it is maintained by the core
julia team and if that breaks on julia-stable, then there will be uproar.

James Fairbanks
PhD Candidate
Georgia Tech

**Comment by sbromberger** [comment]
Yup. I think that's ok.

**Comment by cjamadagni** [comment]
I agree with @jpfairbanks

**Comment by sbromberger** [comment]
Has this been superseded by #277? If so, is it ok to close?

**Comment by cjamadagni** [comment]
This has been superseded by #277. This can be closed.

**Pull request closed without merging**